Alan's work on legend

waharnum opened the pull request:
Creating pull request to incorporate this work into Cindy's FLOE-404 branch.

jobara committed:
FLOE-399: In progress - value entry

Sketched out the data entry page (without styling) in the demos.
Started implemented a valueEntry subcomponent.

jobara committed:
FLOE-399: First pass at valuebinding

Model changes are being output to the DOM, still need to update the model from changes made to inputs in the DOM.

jobara committed:
FLOE-399: Updating the model on change events

jobara committed:
FLOE-399: Calculate percentage

jobara committed:
FLOE-399: Refactoring tests

jobara committed:
FLOE-399: Wired up value binding

jobara committed:
FLOE-399: Renamed valueEntry to dataEntry

jobara committed:
FLOE-399: Fixing classes in dataEntry template

jobara committed:
FLOE-399: Basic rendering of data entry panel

jobara committed:
FLOE-399: Partial commit, moving % calculations

Have started to move the percentage calculations into their own file. Need to make use of these new/refactored methods in dataEntry and dataEntryPanel

jobara committed:
FLOE-399: linting

jobara committed:
FLOE-399: Leaving a TODO comment

jobara committed:
FLOE-399: Using percentage functions

jobara committed:
FLOE-399: render totals based on model

jobara committed:
FLOE-399: Testing totals model value changes

jobara committed:
FLOE-399: Percentage set in the model

jobara committed:
FLOE-399: fixed typo

jobara committed:
FLOE-399: Added method to sum dataEntries

jobara committed:
FLOE-399: Model relay for percentage

Also modified the calculate function to return undefined for invalid entries

jobara committed:
FLOE-399: add containers for dataEntry fields

jobara committed:
FLUID-399: Model relay for dynamic component

jobara committed:
FLOE-399: Updated comment

cindyli committed:
FLOE-399: Fixed file permissions.

jobara committed:
FLOE-399: passing in the containers

Passing in the containers for the dynamic components (dataEntry).

jobara committed:
FLOE-399: rendering dataEntry's template

jobara committed:
FLOE-399: Added a new grade for template injection

jobara committed:
FLOE-399: Updated dataEntry to use injection grade

jobara committed:
FLOE-399: dataEntryPanel with injection grade

Updated the dataEntryPanel component to make use of the templateInjection grade

jobara committed:
FLOE-399: Added dataEntry dynamic components

cindyli committed:
FLOE-399: Implemented the template loading.

jobara committed:
Merge remote-tracking branch 'cindy/FLOE-399' into FLOE-399

jobara committed:
Merge branch 'master' into FLOE-399

jobara committed:
FLOE-399: Passing fetched templates into componets

cindyli committed:
FLOE-399: Add unit tests for gpii.charAuthoring component.

jobara committed:
FLOE-399: Updated class names

jobara committed:
FLOE-399: Linting

jobara committed:
FLOE-399: Fixed empty value cases

jobara committed:
FLOE-399: cleaning up comments

jobara committed:
FLOE-399: clean up of lib files

jobara committed:
FLOE-399: fixing typos

jobara committed:
FLOE-399: refactored test

Refactored the tests for gpii.chartAuthoring.percentage.render to have QUnit perform the DOM cleanup.

jobara committed:
FLOE-399: Removed retrievalVal function

jobara committed:
FLOE-399: added an option for number of digits

The gpii.chartAuthoring.percentage.render method takes in an argument to specify the number of digits after the decimal place to display.

jobara committed:
FLOE-399: Changed description to label

jobara committed:
FLOE-399: Updated test template

jobara committed:
FLOE-399: Data Entry panel only contains the data

removed the chart title, description, and etc.

jobara committed:
FLOE-399: updating model relay

jobara committed:
FLOE-399: Refactoring

Moved more of the logic to model transformations. Keeping the calculations to numbers, and doing any manipulations to strings in rendering functions. Also now returning null in the calculation function instead of undefined, which has a special meaning in the modelTransformation / relay workflow.

jobara committed:
FLOE-399: Implemented a reduce transform

jobara committed:
FLOE-399: Refactoring - renaming, added an option

Renamed setPercnetage to updatePercentage
Added an option for the digits (percentageDigits)

jobara committed:
FLOE-399: Moved invalid tests to a separate run

cindyli commented:
Alan, thanks for the good work.

All commit messages need to start with the JIRA number, such as, 

FLOE-404: ...

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,198 @@
+        },
+        legendOptions: {
+            width: 300,
+            height: 300,
```
"width" and "height" are un-used. They are probably not required for creating legend.

waharnum replied:
Addressed by d49411c

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,198 @@
+            row: "gpii-ca-pieChart-legend-table-row",
+            legendColorCell: "gpii-ca-pieChart-legend-table-row-legend-color-cell",
+            labelCell: "gpii-ca-pieChart-legend-table-row-legend-label-cell",
+            valueCell: "gpii-ca-pieChart-legend-table-row-legend-value-cell"
```
Can we shorten names for cell elements? Same as these element defined in selectors block.

waharnum replied:
Addressed by f8ae24d

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,198 @@
+
+        var table = that.table,
+            l = that.options.legendOptions,
+            dataSet = gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors(that.model.dataSet,that.color),
```
This consolidation should use model relay. An example: https://github.com/fluid-project/infusion/blob/master/tests/framework-tests/core/js/DataBindingTests.js#L947-L955

waharnum replied:
Resolved with 6b021fa

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,198 @@
+        .html(function(d,i) {
+            return d.value;
+        });
+
```
An extra empty line. Same at a few other places.

waharnum replied:
1344ea7

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,198 @@
+        .attr({
+            "class": legendValueCellClass
+        })
+        .html(function(d,i) {
```
The infusion coding convention needs a space in btw the keyword "function" and the open bracket, such as "function (".

The same applies to other spots.

waharnum replied:
d6d9a8a

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,198 @@
+            .attr({
+                "scope":"col"
+            })
+            .html("Legend");
```
The static messages to be shown on UI need to go into "strings" option block so they are user configurable.

The doc: http://docs.fluidproject.org/infusion/development/ComponentConfigurationOptions.html#strings
The example: https://github.com/fluid-project/infusion/blob/master/src/components/overviewPanel/js/OverviewPanel.js#L141

Same comment applies to line 184 and 190.

waharnum replied:
Addressed by f7b712c

cindyli commented on tests/html/legend-Tests.html:
```diff
@@ -0,0 +1,42 @@
+        <script src="../../src/js/legend.js"></script>
+        <script src="../js/legendTests.js"></script>
+        <style>
+        .gpii-ca-pieChart-table td.gpii-ca-pieChart-table td {
```
Is the test dependent on this css? If not, it should be removed. In the case to add the component css to test files, the css should be in a css file in "src/css/" directory.

waharnum replied:
Removed by 19cd02f

cindyli commented on tests/js/legendTests.js:
```diff
@@ -0,0 +1,205 @@
+        gradeNames: ["gpii.chartAuthoring.pieChart.legend", "autoInit"],
+        legendOptions: {
+            width: 200,
+            height: 200,
```
"width" and "height" are un-needed.

cindyli commented on tests/js/legendTests.js:
```diff
@@ -0,0 +1,205 @@
+    };
+
+    jqUnit.test("Test the legend component created based off an array of objects, unsorted", function () {
+        jqUnit.expect(44);
```
The functional flow for testing sorted and unsorted legends are same. Can we consolidate into a function to reuse.

waharnum replied:
They originally had a slightly separate flow, but it became the same on refactoring. Consolidated via 9e700af

cindyli commented:
Some linting errors.

waharnum commented:
I'll look at the other issues later this afternoon, but I've just done a rebase and pushed to put the JIRA number at the start of all the commit messages, since that was easy to address.

jobara committed:
Merge remote-tracking branch 'cindy/FLOE-411'

* cindy/FLOE-411:
  LFOE-411: Upgraded the infusion library for writing unit tests.
  FLOE-411: Upgrade the infusion library to use the latest 2.0 version.

jobara committed:
FLOE-399: Separated extraction and reduction

Separated the extraction and reduction operations into separate functions. These should be passed in independently into the
reduce transform. Also handling the null values in the transformations itself.

jobara committed:
Merge branch 'master' into FLOE-399

jobara committed:
FLOE-399: Updating with Infusion changes

jobara committed:
FLOE-399: linting

waharnum commented:
Looking at the rest of these issues now.

waharnum committed:
FLOE-404: merge Cindy & Alan's work on FLOE-404 together

waharnum committed:
FLOE-404: linting

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,215 @@
+    fluid.defaults("gpii.chartAuthoring.pieChart.legend", {
+        gradeNames: ["gpii.d3ViewComponent", "autoInit"],
+        strings: {
+          legendColHeading:"Legend",
```
Indentations are off here as well as some other spots.

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,215 @@
+            return;
+        }
+
+        that.color = gpii.chartAuthoring.pieChart.legend.getColorScale(colors);
```
This line can be made declarative into the component defaults block.
An example - https://github.com/fluid-project/infusion/blob/master/src/components/textfieldSlider/js/TextfieldSlider.js#L120-L125

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,215 @@
+
+        removedRows.remove();
+
+        if(sort === true) {
```
Can simplify to 

``` javascript
if (sort)
```

Can you add a space btw "if" and the open bracket.

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,215 @@
+      var color = gpii.chartAuthoring.pieChart.legend.getColorScale(model.colors);
+
+        var c = [];
+        for(var i=0; i<dataSet.length; i++) {
```
fluid.each() is encouraged to use for iterations.

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,215 @@
+
+    gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors = function (model) {
+      var dataSet = model.dataSet;
+      var color = gpii.chartAuthoring.pieChart.legend.getColorScale(model.colors);
```
Once this comment https://github.com/cindyli/chartAuthoring/pull/1/files#r38354909 is addressed, "colors" can be directly retrieved form that.colors.

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,215 @@
+        .append("td")
+        .attr({
+            "style": function (d) {
+                return "background-color: "+d.color+";";
```
Can you add spaces before and after "+" sign.

cindyli commented on tests/js/legendTests.js:
```diff
@@ -0,0 +1,192 @@
+
+    gpii.tests.chartAuthoring.mouseOverListener = function (data, i, that) {
+        that.mouseOverListenerCalled = true;
+    };
```
Extra spaces in front line 90 - 92.

cindyli commented on tests/js/legendTests.js:
```diff
@@ -0,0 +1,192 @@
+      // Item removed from dataset
+
+      that.applier.change("dataSet", gpii.tests.chartAuthoring.objectArrayRemove);
+      gpii.tests.chartAuthoring.validateLegend(that, false);
```
gpii.tests.chartAuthoring.validateLegend.validateLegend() doesn't accept the second argument.

waharnum committed:
FLOE-404: clean up some incorrect indents

waharnum committed:
FLOE-404: use fluid.each instead of for loop in consolidateDataAndColors function; make creation of color scale declarative in default block

waharnum committed:
Simplify sort test

waharnum committed:
Better concatentation readability

waharnum commented:
Comments should be addressed now.

cindyli commented:
Thanks for the fix, @waharnum. They look great. 

BTW, there are some linting errors.

cindyli commented on src/js/legend.js:
```diff
@@ -0,0 +1,222 @@
+            labelColHeading:"Label",
+            valueColHeading:"Value"
+        },
+        model: {
```
Would be nice to add a comment for "dataSetWithColors" path in the model block. It makes easier for other developers to realize as well as for the documentation purpose.

cindyli replied:
Sorry I didn't make it clear. What I meant of adding a comment for "dataSetWithColors" was,

``` javascript
model: {
     dataSet: [],
    // "dataSetWithColors" consolidates the dataset in original order with the array of colors, into a single array of objects
    // dataSetWithColors: []
}
```

When writing component API docs, we normally look at this defaults block. This comment helps to realize there's another model path that's created internally.

waharnum replied:
Gotcha, correcting now.

waharnum committed:
FLOE-404: linting

waharnum committed:
FLOE-404: add comments on modelRelay intention

waharnum committed:
FLOE-404: make appearance of column headings in legend configurable

waharnum committed:
FLOE-404: merge jobara's work from FLOE-399

waharnum committed:
FLOE-404: fix an issue with updating values and labels when data changes in place, as opposed to being added or removed

waharnum committed:
FLOE-404: Add additional logic to prevent concatenation of style+selector when their class names are identical

waharnum committed:
FLOE-404: Add test of sync between displayed numeric values and data model

waharnum committed:
FLOE-404: refine name of sync test function

waharnum committed:
FLOE-404: missing semicolon

waharnum committed:
FLOE-404: test for in-place update of legend labels and values

waharnum committed:
FLOE-404: linting after downgrading JSHint to 2.4.0 for indentation warnings

waharnum committed:
FLOE-404: add test to check synthesization does not double up class names when selector & style are the same

waharnum committed:
FLOE-404: Refactor gpii.d3ViewComponent.synthesizeClasses to handle some other cases & ensure uniqueness of synthesized classes

waharnum committed:
FLOE-404: Add tests for identical styles/selectors, and mix of overlapping & unique styles/selectors

waharnum committed:
FLOE-404: complicated style/selector names to include hyphens and underscores (legal in CSS classes)

waharnum committed:
FLOE-404: use fluid.set instead of direct interaction with consolidatedClasses object

cindyli commented:
@waharnum, it seems you merged in the branch for the data entry form and committed. Can you clean it up as this branch is only for D3 works. Thanks.

cindyli commented on src/js/d3ViewComponent.js:
```diff
@@ -56,26 +56,46 @@ https://github.com/gpii/universal/LICENSE.txt
+        fluid.each(consolidatedClasses, function(classes, key) {
+            var splitClasses = classes.split(" ");
+            // Basic alphabetic sort to improve readability of applied classes and make testing easier
+            splitClasses.sort(function (a,b) {
```
Infusion's convention for applying selectors to DOM elements are, control selector names, the ones starting with "gpiic-" in the chart authoring context, comes before styling selector names that start with "gpii-". The convention was decided 'cuz usually a DOM element only has one control selector but several styling selectors. Would be nice it the consolidated classes can sustain this convention.

cindyli commented on src/js/pie.js:
```diff
@@ -126,6 +123,12 @@ https://github.com/gpii/universal/LICENSE.txt
             });
 
+        // Update text values
+        texts.text(function (d) {
```
Thanks for fixing this. A beautiful finding.

cindyli commented on tests/js/legendTests.js:
```diff
@@ -0,0 +1,232 @@
+
+    gpii.tests.chartAuthoring.testLegendSyncWithModelDataSet = function(that) {
+        var dataSet = that.model.dataSet;
+        //
```
Stray comment signs.

waharnum committed:
Revert "FLOE-404: merge jobara's work from FLOE-399"

This reverts commit 3753af1f6811dba20bce9b6dfcf061065f1ea307, reversing
changes made to b3df97eac7a358367ffcd0d2a1682686e45502e6.

cindyli commented on tests/js/legendTests.js:
```diff
@@ -0,0 +1,232 @@
+        var dataSet = that.model.dataSet;
+        //
+        if(that.options.legendOptions.sort) {
+            dataSet.sort(function(a,b) {
```
Exactly same sorting function used in legend.js. If this piece of code is wrong, the test is not catching it. You probably wanna directly pass in the sorted array value for comparison.

waharnum replied:
It's actually that we need to sort the model's dataSet in test in ascending order to compare it against the rendered components, if the sort option is true. At the moment, legend.js leaves the original dataset's ordering alone, but sorts the D3-based elements when sort: true.

It is possible this is overly complicated & we should simply sort the dataSet whenever it's updated if the sort option is true in legend.js. I think my logic was that you may want to preserve original dataset order for other reasons while sorting the visual display.

waharnum committed:
FLOE-404: remove orphaned comma

cindyli commented on tests/js/pieTests.js:
```diff
@@ -133,8 +146,19 @@ https://github.com/gpii/universal/LICENSE.txt
         value: 45
     }];
 
+    gpii.tests.chartAuthoring.objectArrayChangeInPlace = [{
```
Nice catch. Thanks for adding this test scenario.

waharnum committed:
FLOE-404: synthesizeClasses now does not sort & processes selectors first, to maintain gpiic/gpii ordering convention; add tests to cover this

waharnum committed:
FLOE-404: add dataSetWithColor[] comment to default model, with explanation

cindyli commented:
There are some linting errors.

waharnum committed:
FLOE-404: add sorted object arrays as fixed test data, for better testing of model syncing with displayd values

waharnum committed:
FLOE-404: linting
